File: solutions/csharp/custom-set/1/CustomSet.cs
// Solution respectfully dedicated to Shree DR.MDD

public class CustomSet(params int[] values)
{
    protected List<int> coreSet = new List<int>(values);

    public CustomSet Add(int item) =>
        Exists(item)
        ? new CustomSet(coreSet.ToArray())
        : new CustomSet(coreSet.ToArray().Append(item).ToArray());

    public bool Empty() => coreSet.Count == 0;

    public bool Contains(int item) => coreSet.Contains(item);

    public bool Subset(CustomSet other) => !coreSet.Any(elem => !other.Contains(elem));

    public bool Disjoint(CustomSet other) => !coreSet.Any(elem => other.Contains(elem));

    public CustomSet Intersection(CustomSet other) =>
        new CustomSet(coreSet.Where(elem => other.Contains(elem)).ToArray());

    public CustomSet Difference(CustomSet other) =>
        new CustomSet(coreSet.Where(elem => !other.Contains(elem)).ToArray());

    public CustomSet Union(CustomSet other) =>
        new CustomSet(coreSet.Union(other.coreSet).Distinct().ToArray());

    public override bool Equals(object obj) => Equals(obj as CustomSet);

    public bool Equals(CustomSet other)
    {
        if (other == null) return false;

        var firstSorted = coreSet.Distinct().Order().ToArray();
        var secondSorted = other.coreSet.Distinct().Order().ToArray();

        if (firstSorted.Length != secondSorted.Length)
            return false;

        for (int i = 0; i < firstSorted.Length; i++)
        {
            if (firstSorted[i] != secondSorted[i])
                return false;
        }

        return true;
    }

    public override int GetHashCode() =>
        coreSet.Aggregate(0, (acc, val) => unchecked(acc * 31 + val));

    private bool Exists(int key) => coreSet.Contains(key);
}

File: solutions/python/custom-set/1/custom_set.py
# Dedicated to Shree DR.MDD 🙏

from __future__ import annotations


class CustomSet:
    def __init__(self, elements=None):
        self._bucket = list(elements or [])

    def isempty(self) -> bool:
        return not self._bucket

    def __contains__(self, item) -> bool:
        return item in self._bucket

    def issubset(self, other) -> bool:
        return all(e in other for e in self)

    def isdisjoint(self, other) -> bool:
        return not any(e in self for e in other)

    def __eq__(self, other) -> bool:
        return self.issubset(other) and other.issubset(self)

    def add(self, item) -> None:
        if item not in self._bucket:
            self._bucket.append(item)

    def intersection(self, other) -> CustomSet:
        return CustomSet(x for x in other if x in self)

    def __sub__(self, other) -> CustomSet:
        return CustomSet(x for x in self if x not in other)

    def __add__(self, other) -> CustomSet:
        merged = CustomSet(self)
        for val in other:
            merged.add(val)
        return merged

    def __iter__(self):
        return iter(self._bucket)
